sklearn/semi_supervised/self_training.py

```diff
@@ -0,0 +1,239 @@
+        """
+        X, y = check_X_y(X, y)
+        _check_estimator(self.base_estimator)
+        self.base_estimator_ = clone(self.base_estimator)
+
+        if not 0 <= self.max_iter:
+            raise ValueError("max_iter must be >= 0")
+
+        if not 0 <= self.threshold < 1:
+            raise ValueError("threshold must be in [0,1)")
+
+        # Data usable for supervised training
+        self.X_labeled_ = X[safe_mask(X, np.where(y != -1))][0]
+        self.y_labeled_ = y[safe_mask(y, np.where(y != -1))][0]
+        self.y_labeled_iter_ = np.full_like(self.y_labeled_, 0)
+
+        # Unlabeled data
+        X_unlabeled = X[safe_mask(X, np.where(y == -1))][0]
+        y_unlabeled = y[safe_mask(y, np.where(y == -1))][0]
+
+        iter = 0
+        while len(self.X_labeled_) < len(X) and iter < self.max_iter:
+            iter += 1
+            self.base_estimator_.fit(self.X_labeled_, self.y_labeled_)
+
+            # Select predictions where confidence is above the threshold
+            predict_proba = self.base_estimator_.predict_proba(X_unlabeled)
+
+            pred = np.argmax(predict_proba, axis=1)
+            max_proba = np.max(predict_proba, axis=1)
+
+            confident = np.where(max_proba > self.threshold)[0]
+
+            # Add newly labeled confident predictions to the dataset
+            self.X_labeled_ = np.append(
+                self.X_labeled_, X_unlabeled[confident], axis=0)
+            self.y_labeled_ = np.append(
+                self.y_labeled_, pred[confident], axis=0)
+            self.y_labeled_iter_ = np.append(
+                self.y_labeled_iter_, np.full_like(confident, iter))
+
+            # Remove already labeled data from unlabeled dataset
+            X_unlabeled = np.delete(X_unlabeled, confident, axis=0)
+            y_unlabeled = np.delete(y_unlabeled, confident, axis=0)
+
+        self.base_estimator_.fit(self.X_labeled_, self.y_labeled_)
+        return self
+
+    @if_delegate_has_method(delegate='base_estimator')
+    def predict(self, X):
+        """Predict on a dataset.
+
+        Parameters
+        ----------
+        X : array-like, shape = (n_samples, n_features)
+            array representing the data
+
+        Returns
+        -------
+        y : array-like, shape = (n_samples, 1)
+            array with predicted labels
+        """
+        check_is_fitted(self, 'y_labeled_iter_')
+        X = check_array(X)
+        return self.base_estimator_.predict(X)
+
+    def predict_proba(self, X):
+        """Predict probability for each possible outcome.
+
+        Parameters
+        ----------
+        X : array-like, shape = (n_samples, n_features)
+            array representing the data
+
+        Returns
+        -------
+        y : array-like, shape = (n_samples, n_features)
+            array with prediction probabilities
+        """
+        check_is_fitted(self, 'y_labeled_iter_')
+        return self.base_estimator_.predict_proba(X)
```

Comment:
This line is for some reason not run by tests

Reply:
Fixed.